### PR TITLE
Multi-component sparsity pattern

### PR DESCRIPTION
**Pull request by chleh**
This PR provides a multi-component aware implementation for setting matrix sparsities.

### CODE REVIEW

**Comment by TomFischer** [comment]
Nice! :+1:

File: AssemblerLib/ComputeSparsityPattern.cpp [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+#include "ComputeSparsityPattern.h"
+
+namespace AssemblerLib
+{
+
+SparsityPattern
+computeSparsityPattern(LocalToGlobalIndexMap const& dof_table,
+        MeshLib::Mesh const& mesh
+        )
+{
```
Comment (wenqing) [inline AssemblerLib/ComputeSparsityPattern.cpp]: It seems that this sparsity pattern is only for the case of indexing the entries of the stiffness matrix by component. If the global indexing is performed by location, the  sparsity pattern is made up of a number of same patches that are determined by NodeAdjacencyTable solely.

Comment (chleh) [inline AssemblerLib/ComputeSparsityPattern.cpp]: Question resolved in discussion with @wenqing.

Comment (wenqing) [inline AssemblerLib/ComputeSparsityPattern.cpp]: Question is resolved after discussion with @chleh .

**Comment by wenqing** [comment]
:+1:

File: AssemblerLib/ComputeSparsityPattern.cpp [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        for (auto an : node_ids) {
+            auto const& row_ids = global_idcs[an];
+            for (auto r : row_ids) {
+                sparsity_pattern[r] += row_ids.size();
+            }
+        }
+    }
+
+    return sparsity_pattern;
+}
```
Comment (endJunction) [inline AssemblerLib/ComputeSparsityPattern.cpp]: Could you add a few comments explaining main steps of the algorithm?

File: AssemblerLib/ComputeSparsityPattern.h [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+#include <vector>
+
+#include "ProcessLib/NumericsConfig.h"
+
+namespace AssemblerLib
+{
+
+class LocalToGlobalIndexMap;
+
+using SparsityPattern = std::vector<GlobalIndexType>;
```
Comment (endJunction) [inline AssemblerLib/ComputeSparsityPattern.h]: Comment the purpose of the type something along the lines:
```c++
/// Vector of number of nonzero entries in global matrix.
```
and change the return description of `computeSparsityPattern()` below---now it is returning a `SparsityPattern`, not a vector.

File: AssemblerLib/ComputeSparsityPattern.cpp (hunk 2 above)
Comment (chleh) [inline AssemblerLib/ComputeSparsityPattern.cpp]: Comments added.

**Pull request merged**